pkg/workflow/workflow.go

```diff
@@ -351,22 +351,34 @@ func (w *workflow) setMetadataToContext(wfCtx wfContext.Context) error {
 	return wfCtx.SetVar(metadata, wfTypes.ContextKeyMetadata)
 }
 
+func (e *engine) getMinTimes(stepID string, minTimes *int) (found bool) {
```

Comment:
I don't think it's good to pass a int pointer to get the minTimes, you can:
```go
func (e *engine) getMinTimes(stepID string) int
```

If not found, you can return `0`.

Reply:
If I understood correctly, it means get how much times left here.

```suggestion
func (e *engine) getLeftChances(stepID string, leftTimes *int) int {
```